File: chistmas-caching/reading/src/cache-manager.js
const CACHE_LAYER = new Map()

class CacheManager {
    /**
     * Gets cached entry associated with key, Returns null if missing.
     * 
     * @param {string} key 
     * @returns {any|null}
     */
    get(key) {
        if (!CACHE_LAYER.has(key)) {
            return null
        }

        return CACHE_LAYER.get(key)
    }

    /**
     * Saves a key in cache
     * 
     * @param {string} key 
     * @param {any} entry 
     * @param {number} [expiration=0] - in seconds 
     */
    set(key, entry, expiration = 0) {
        if (!key) {
            throw new RangeError(`Cannot set entry without a key. Expected a key, got ${key}`)
        }

        CACHE_LAYER.set(key, entry)

        if (expiration) {
            setTimeout(() => {
                CACHE_LAYER.delete(key)
            }, expiration * 1000)
        }
    }
}

module.exports = CacheManager

File: chistmas-caching/writing/src/cache_manager.py
from threading import Timer
from utils import trace

class CacheManager:
    def __init__(self):
        self.cache = {}

    @trace
    def set(self, key, value, expiration=0):
        if not key:
            raise ReferenceError("Cannot set entry without a key. Expected a key, got None" + str(key))

        self.cache[key] = value

        if expiration:
            timeout = Timer(expiration, self.__delete, [key])
            timeout.start()
    
    @trace
    def get(self, key):
        if not key:
            raise ReferenceError("Cannot get entry without a key. Expected a key, got " + str(key))

        if key not in self.cache.keys():
            return None

        return self.cache[key]
    
    def __delete(self, key):
        if not key:
            raise ReferenceError("Cannot delete entry without a key. Expected a key, got " + str(key))
        
        del self.cache[key]